### PR TITLE
Add folder requests.

### PR DESCRIPTION
**Pull request by olha-khomiakova**
Implement the ability to send requests to the server to get the user preferences, the data of a specific folder, the content of a specific folder.
All tests passed.
ESLint passed.

### CODE REVIEW

File: js/app/services/api-service/index.js [hunk 1]
```diff
@@ -51,24 +53,57 @@ export class ApiService {
...
+  getFolder(id) {
+    return this._window.fetch(`/folder/${id}`, {
+      method: 'GET',
+      headers: {
+        'content-type': 'application/json',
+      },
+    }).then((response) => {
+      return this._checkResponse(response);
+    }).then((json) => {
+      return Promise.resolve(new FolderListModel(json));
```
Comment (dkalinin) [inline js/app/services/api-service/index.js:86]: It is not necessary return a promise from `then`. You can return any value. If it returns promise then the next `then` will receive a result of this `then`.

File: js/app/services/api-service/index.js [hunk 2]
```diff
@@ -51,24 +53,57 @@ export class ApiService {
...
+  }
+
+
+  /**
+   * Sends a request to the server to get the folder content by the folder ID.
+   * @param {String} id stores information about the folder unique identifier
+   * @returns {Promise<ListItemModel>}
+   */
+  getFolderContent(id) {
+    return this._window.fetch(`/folder/00000/content`, {
```
Comment (dkalinin) [inline js/app/services/api-service/index.js:97]: `id` should be used instead of `000000`.

File: js/app/services/api-service/index.js [hunk 3]
```diff
@@ -86,12 +121,14 @@ export class ApiService {
       throw new ServerErrorResponse(serverErrorMessage);
     } else if (response.status >= 400) {
       if (response.status === 422) {
-        throw new ValidationErrorResponse(response.body.errors);
+        return response.json().then((responseJson) => {
+          throw new ValidationErrorResponse(responseJson);
+        });
       } else if (response.status === 403) {
         throw new ClientErrorResponse('Authentication error.');
       }
```
Comment (dkalinin) [inline js/app/services/api-service/index.js:129]: What if status code is 400?

File: js/app/file-list-page/file-list-page/index.js [hunk 4]
```diff
@@ -61,23 +57,44 @@ export class FileListPage extends Component {
...
+
+    this.apiService.userPreferences().then((userPreferences) => {
+      this._userPreferences = userPreferences;
+      userDetails.username = this._userPreferences.name;
+
+      this.apiService.getFolder(this._userPreferences.rootFolderId).then((folder) => {
+        const breadcrumbs = new Breadcrumbs(this.getElement('breadcrumbs'));
+        breadcrumbs.addBreadcrumb(folder.name);
+      }).catch((error) => {
+        this._serverError = error.message;
```
Comment (dkalinin) [inline js/app/file-list-page/file-list-page/index.js:74]: Breadcrumbs component should show error state.

File: js/app/file-list-page/file-list-page/index.js [hunk 5]
```diff
@@ -61,23 +57,44 @@ export class FileListPage extends Component {
...
+      });
+
+      this.apiService.getFolderContent(this._userPreferences.rootFolderId).then((list) => {
+        const fileList = new FileList(this.getElement('file-list'));
+        fileList.addList(list);
+      }).catch((error) => {
+        this._serverError = error.message;
```
Comment (dkalinin) [inline js/app/file-list-page/file-list-page/index.js:81]: FileList component should show error state.

File: js/app/file-list-page/file-list-page/index.js [hunk 6]
```diff
@@ -61,23 +57,44 @@ export class FileListPage extends Component {
...
+      });
+    }).catch((error) => {
+      this._serverError = error.message;
```
Comment (dkalinin) [inline js/app/file-list-page/file-list-page/index.js:84]: Message should say that it was unable to load user preferences.

**Review by dkalinin — CHANGES_REQUESTED** [review]
PTAL

File: js/app/services/api-service/index.js (hunk 1 above)
Comment (olha-khomiakova) [reply js/app/services/api-service/index.js:86]: Done.

File: js/app/services/api-service/index.js (hunk 2 above)
Comment (olha-khomiakova) [reply js/app/services/api-service/index.js:97]: Done.

File: js/app/services/api-service/index.js (hunk 3 above)
Comment (olha-khomiakova) [reply js/app/services/api-service/index.js:129]: Done.

File: js/app/file-list-page/file-list-page/index.js (hunk 5 above)
Comment (olha-khomiakova) [reply js/app/file-list-page/file-list-page/index.js:81]: Done.

File: js/app/file-list-page/file-list-page/index.js (hunk 4 above)
Comment (olha-khomiakova) [reply js/app/file-list-page/file-list-page/index.js:74]: Done.

File: js/app/file-list-page/file-list-page/index.js (hunk 6 above)
Comment (olha-khomiakova) [reply js/app/file-list-page/file-list-page/index.js:84]: Done.

File: js/app/file-list-page/breadcrumbs/index.js [hunk 7]
```diff
@@ -15,6 +15,13 @@ export class Breadcrumbs extends Component {
     this.getElement('crumbs').innerText += '/ ' + value;
   }
 
+  /**
+   * Sets an error massage.
```
Comment (dkalinin) [inline js/app/file-list-page/breadcrumbs/index.js:19]: M*a*ssage is good, but m*e*ssage fits better.

File: js/app/services/api-service/index.js [hunk 8]
```diff
@@ -94,41 +94,54 @@ export class ApiService {
...
 
 
   /**
    * Checks the response status.
-   * Returns Promise.resolve(responseJson) if the response status is 200
+   * Returns responseJson if the response status is 200
    * otherwise throws an error.
    * @param {Response} response
-   * @returns {Promise<ItemJson|ItemJson[]|Promise>}
+   * @param {Object} errorMessages
```
Comment (dkalinin) [inline js/app/services/api-service/index.js:115]: Please add description. Also it's better to specify possible keys.

File: js/app/file-list-page/file-list-page/index.js [hunk 9]
```diff
@@ -61,27 +61,30 @@ export class FileListPage extends Component {
...
+      this._userDetails.username = userPreferences.name;
 
-      this.apiService.getFolder(this._userPreferences.rootFolderId).then((folder) => {
-        const breadcrumbs = new Breadcrumbs(this.getElement('breadcrumbs'));
-        breadcrumbs.addBreadcrumb(folder.name);
+      this.apiService.getFolder(userPreferences.rootFolderId).then((folder) => {
+        this._breadcrumbs.addBreadcrumb(folder.name);
       }).catch((error) => {
-        this._serverError = error.message;
+        this._breadcrumbs.setErrorMessage();
```
Comment (dkalinin) [inline js/app/file-list-page/file-list-page/index.js:75]: Why don't we pass message here?

File: js/app/file-list-page/file-list-page/index.js [hunk 10]
```diff
@@ -61,27 +61,30 @@ export class FileListPage extends Component {
...
       });
 
-      this.apiService.getFolderContent(this._userPreferences.rootFolderId).then((list) => {
-        const fileList = new FileList(this.getElement('file-list'));
-        fileList.addList(list);
+      this.apiService.getFolderContent(userPreferences.rootFolderId).then((list) => {
+        this._fileList.addList(list);
       }).catch((error) => {
-        this._serverError = error.message;
+        this._fileList.setErrorMessage();
```
Comment (dkalinin) [inline js/app/file-list-page/file-list-page/index.js:81]: Why don't we pass message here?

File: js/app/file-list-page/file-list-page/index.js [hunk 11]
```diff
@@ -61,27 +61,30 @@ export class FileListPage extends Component {
...
       });
     }).catch((error) => {
       this._serverError = error.message;
+      this._userDetails.setErrorMessage();
+      this._breadcrumbs.setErrorMessage();
+      this._fileList.setErrorMessage();
```
Comment (dkalinin) [inline js/app/file-list-page/file-list-page/index.js:87]: One error message is enough here.

**Review by dkalinin — APPROVED** [review]
Please add dev-fetch logging and tests to check hoe error message is rendered for each API call.

File: js/app/file-list-page/breadcrumbs/index.js (hunk 7 above)
Comment (olha-khomiakova) [reply js/app/file-list-page/breadcrumbs/index.js:19]: Done.

File: js/app/services/api-service/index.js (hunk 8 above)
Comment (olha-khomiakova) [reply js/app/services/api-service/index.js:115]: Done.

File: js/app/file-list-page/file-list-page/index.js (hunk 9 above)
Comment (olha-khomiakova) [reply js/app/file-list-page/file-list-page/index.js:75]: Done.

File: js/app/file-list-page/file-list-page/index.js (hunk 10 above)
Comment (olha-khomiakova) [reply js/app/file-list-page/file-list-page/index.js:81]: Done.

File: js/app/file-list-page/file-list-page/index.js (hunk 11 above)
Comment (olha-khomiakova) [reply js/app/file-list-page/file-list-page/index.js:87]: Done.

**Pull request merged**